File: 874-backspace-string-compare/backspace-string-compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        return readyStr(s).equals(readyStr(t));
    }

    private String readyStr(String str) {
        // approach 1 - using stack
        /*
        Stack<Character> stack = new Stack<>();

        for (int i=0; i<str.length();i++) {
            char curr = str.charAt(i);

            if (curr!='#') stack.push(curr);
            else if (!stack.isEmpty()) stack.pop();
        }
        // return stack.join('');
        return stack.toString();
        */

        //approach 2 - using Map (cool and easy)...
        Map<Integer, Character> map = new HashMap<>();
        int orderKey = 0;

        for (int i=0; i<str.length();i++) {
            char curr = str.charAt(i);

            if (curr!='#') {
                map.put(orderKey,curr);
                orderKey++;
            } else if (orderKey > 0) {
                orderKey--;
                map.remove(orderKey);
            }
        }

        // return [...map.values()].join('');
        return new ArrayList<>(map.values()).toString();
    }
}

File: 874-backspace-string-compare/backspace-string-compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var backspaceCompare = function(s, t) {

    const readyStr = (str) => {
        // approach 1 - using stack
        // let stack = [];

        // for(let i=0;i<str.length;i++) {
        //     let curr = str[i];

        //     if(curr!='#') stack.push(curr);
        //     else if(stack.length>0) stack.pop();
        // }
        // return stack.join('');


        //approach 2 - using Map (cool and easy)...
        const map = new Map();
        let orderKey = 0;

        for(let i=0;i<str.length;i++) {
            let curr = str[i];

            if(curr!='#'){

                map.set(orderKey,curr);
                orderKey++;
            }else {
                orderKey--;
                map.delete(orderKey)
            }
        }

        return [...map.values()].join('');

    }
    //writing the func onyl 1 time bcos both need same working ...jsut giving the ip params and ending ...
    return readyStr(s) === readyStr(t);
};